### PR TITLE
Add Reactive retries for SC LoadBalancer

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
Fixes gh-659.

### CODE REVIEW

File: docs/src/main/asciidoc/_configprops.adoc [hunk 1]
```diff
@@ -32,6 +32,10 @@
 |spring.cloud.loadbalancer.health-check.path |  | 
 |spring.cloud.loadbalancer.hint |  | Allows setting the value of <code>hint</code> that is passed on to the LoadBalancer request and can subsequently be used in {@link ReactiveLoadBalancer} implementations.
 |spring.cloud.loadbalancer.retry.avoid-previous-instance | true | Enables wrapping ServiceInstanceListSupplier beans with `RetryAwareServiceInstanceListSupplier` if Spring-Retry is in the classpath.
+|spring.cloud.loadbalancer.retry.backoff.enabled | false | Indicates whether Reactor Retry backoffs should be applied.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/_configprops.adoc:35]: Add backticks around `false`, to make it code.

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 2]
```diff
@@ -437,18 +437,27 @@ Then, `ReactiveLoadBalancer` is used underneath.
...
+For the reactive implementation, you just need to enable it by setting `spring.cloud.loadbalancer.retry.backoff.enabled` to `false`.
 
 You can set:
 
 - `spring.cloud.loadbalancer.retry.maxRetriesOnSameServiceInstance` - indicates how many times a request should be retried on the same `ServiceInstance` (counted separately for every selected instance)
 - `spring.cloud.loadbalancer.retry.maxRetriesOnNextServiceInstance` - indicates how many times a request should be retried a newly selected `ServiceInstance`
 - `spring.cloud.loadbalancer.retry.retryableStatusCodes` - the status codes on which to always retry a failed request.
 
+For the reactive implementation, you can additionally set:
+ - `spring.cloud.loadbalancer.retry.backoff.minBackoff` - sets minimum backoff duration (by default, 5 milliseconds)
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:455]: Capitalize "sets" and "the" between "sets" and "minimum". So it becomes "- Sets the minimum...."

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 3]
```diff
@@ -437,18 +437,27 @@ Then, `ReactiveLoadBalancer` is used underneath.
...
+ - `spring.cloud.loadbalancer.retry.backoff.maxBackoff` - sets maximum backoff duration (by default, max long value of milliseconds)
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:456]: Capitalize "sets" and "the" between "sets" and "minimum". So it becomes "- Sets the maximum...."

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 4]
```diff
@@ -437,18 +437,27 @@ Then, `ReactiveLoadBalancer` is used underneath.
...
+ - `spring.cloud.loadbalancer.retry.backoff.jitter` - sets the jitter used for calculationg the actual backoff duration for each call (by default, 0.5).
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:457]: Capitalize "sets"

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 5]
```diff
@@ -437,18 +437,27 @@ Then, `ReactiveLoadBalancer` is used underneath.
...
+
+For the reactive implementation, you can also implement your own `LoadBalancerRetryPolicy` to have a more detailed control over the load-balanced call retries.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:459]: Change "have a more detailed" to "have more detailed" (remove "a").

**Review by Buzzardo — CHANGES_REQUESTED** [review]
I always hit a bunch of details, but, really, you write well.  :)

File: docs/src/main/asciidoc/_configprops.adoc (hunk 1 above)
Comment (spencergibb) [reply docs/src/main/asciidoc/_configprops.adoc:35]: this is auto-generated.

**Review by marcingrzejszczak — APPROVED** [review]

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ExchangeFilterFunctionTools.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.web.reactive.function.client.ClientRequest;
+import org.springframework.web.reactive.function.client.ExchangeFilterFunction;
+
+/**
+ * A utility class for load-balanced {@link ExchangeFilterFunction} instances.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 3.0.0
+ */
+public final class ExchangeFilterFunctionTools {
```
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ExchangeFilterFunctionTools.java:31]: Possibly. Let me change that.

**Pull request merged**